### PR TITLE
[MLP-76] Support CRUD operations using mongodb

### PR DESCRIPTION
**Pull request by matiasbeckerle**
## Background

Hey folks. I played a little bit with mongoose in order to support the basic CRUD operations using User as an example.

## Changes

- Added User schema & model.
- Added common mongoose plugin to override `toJSON` default behavior and remove the private & internal `_id` field.
- Updated CRUD operations to use mongodb instead of memory.
- Updated some components because ids are not numbers anymore.

### CODE REVIEW

File: api/database/mongooseIdPlugin.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export default function mongooseIdPlugin(schema, options) {
```
Comment (matiasbeckerle) [inline api/database/mongooseIdPlugin.js]: I don't like the name.

File: api/database/mongooseIdPlugin.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  schema.options.toJSON.transform = function (doc, ret, options) {
+    delete ret._id; // Remove internal _id property before exposing it as JSON.
```
Comment (AlphaGit) [inline api/database/mongooseIdPlugin.js]: Why would you want to remove the `_id` property?

**Review by AlphaGit — APPROVED** [review]
Looks good so far!

File: api/database/mongooseIdPlugin.js (hunk 2 above)
Comment (matiasbeckerle) [inline api/database/mongooseIdPlugin.js]: Good question. I'm new to mongodb & mongoose but AFAIK `_id` is an internal property they use to identify a document which is absolutely fine. They also provide a getter for that property so when you use `product.id` it automatically returns `_id` which is great, right? Well... I want to use that property for everything except for exposing it. So, whatever API endpoint that retrieves a document, it will automatically remove the `_id` but provide the `id` field.

TL;DR: wherever we ask for a product it will return `{ id: 'some-random-id', name: 'awesome product' }` instead of `{ _id: 'some-random-id', name: 'awesome product', id: 'some-random-id' }`.

Comment (AlphaGit) [inline api/database/mongooseIdPlugin.js]: Sounds good. If the JSON object can then be used seamlessly to update the document (with `id` instead of `_id`), then we should be fine.

But generally, I think we should better avoid having transformations in and out of the database, since it could create "mapping issues".

Comment (matiasbeckerle) [inline api/database/mongooseIdPlugin.js]: > But generally, I think we should better avoid having transformations in and out of the database, since it could create "mapping issues".

Totally agree with that.

But. Regarding your first sentence, I want to clarify that `update` is based on the id provided. So it is handled manually. If you can, could you take a quick look in that method? I believe is pretty straightforward but is also discussable.

In certain way I believe what I achieved is robust based on the mongodb & mongoose behavior but maybe someone with more experience have a nicer alternative. Thoughts, ideas?

Comment (AlphaGit) [inline api/database/mongooseIdPlugin.js]: I actually think that implementation looks fine. Sure, we could discuss a lot about validation and security, but I think that this is a trimmed down version of a real system (that is meant to be created from it), so what concerns you exactly?

Comment (matiasbeckerle) [inline api/database/mongooseIdPlugin.js]: Exactly. As you said, is a trimmed down version and for that reason I believe this implementation is pretty solid for now. In much larger systems, `_id` could be re-evaluated.

**Comment by matiasbeckerle** [comment]
Updated code to ensure mongodb is connected.

**Review by mravinale — APPROVED** [review]

**Review by AlphaGit — APPROVED** [review]

**Review by mravinale — APPROVED** [review]

File: cli/generators/crud/index.js [hunk 3]
```diff
@@ -7,6 +7,13 @@ module.exports = class extends Generator {
     super(args, opts);
 
     this.argument('name', { type: String, required: true });
+
+    this.option('api', {
+      desc: 'Does not include API related files',
```
Comment (AlphaGit) [inline cli/generators/crud/index.js]: Is this an indicator of what the option does? If so, would it be better that it indicates it in imperative? ("Do not include API related files")

File: cli/generators/crud/templates/service.js [hunk 4]
```diff
@@ -1,67 +1,43 @@
...
-    let <%= name %> = <%= pluralizedName %>.find(<%= name %> => <%= name %>.id == id);
-    Object.assign(<%= name %>, data);
-
-    // Put the updated record on memory.
-    <%= pluralizedName %>.splice(indexOf<%= ucName %>ToUpdate, 1, <%= name %>);
-
-    return cb(null, <%= name %>);
+    delete data.id;
+
+    <%= ucName %>.findByIdAndUpdate(id, data, { new: true }) // To return the modified document rather than the original.
```
Comment (AlphaGit) [inline cli/generators/crud/templates/service.js]: Super-ultra minor: do you think this comment would be better in the line above?

File: cli/test/crud.js [hunk 5]
```diff
@@ -16,6 +16,11 @@ describe('generator-mern:crud', function () {
       .toPromise();
   });
 
+  it('creates an schema file', function () {
```
Comment (AlphaGit) [inline cli/test/crud.js]: Minor (grammar): "Creates **a** schema file"

**Review by AlphaGit — APPROVED** [review]

File: cli/generators/crud/index.js (hunk 3 above)
Comment (matiasbeckerle) [inline cli/generators/crud/index.js]: Crap. This commit is not supposed to be here. I didn't create a new branch. Will do it now.

Regarding your suggestion, yes, I like it better!

File: cli/generators/crud/templates/service.js (hunk 4 above)
Comment (matiasbeckerle) [inline cli/generators/crud/templates/service.js]: No, and I will explain why. The comment is only referred to `new: true` param which is... waaay to unspecific right? I use comments in the line above when is something more important or related to the overall line. This case is like explaining something related to one of the params. When is just a minor warning I prefer to leave it on the right side. Anyways is a case-by-case decision.

But... if you think the comment is important enough maybe I should put it above. Now I have doubts 😄

File: cli/test/crud.js (hunk 5 above)
Comment (matiasbeckerle) [inline cli/test/crud.js]: Thanks!

File: cli/generators/crud/templates/service.js (hunk 4 above)
Comment (AlphaGit) [inline cli/generators/crud/templates/service.js]: No, I don't think it's that important, but since it's way to the right, it's easy to miss. Check this out:

image

Regardless, do you think you could indicate what you were referring to, since I don't think it's obvious. (`// using { new: true } to return the modified value... `

Also, doesn't `new: true` have the opposite effect to what you described?

Comment (matiasbeckerle) [inline cli/generators/crud/templates/service.js]: Like your proposal.

Unfortunately, no. It causes what is described. "True" in this particular API means that it will return the updated value. Is a really really confusing signature. http://mongoosejs.com/docs/api.html#model_Model.findByIdAndUpdate

File: cli/generators/crud/templates/service.js [hunk 6]
```diff
@@ -1,67 +1,43 @@
...
-    let <%= name %> = <%= pluralizedName %>.find(<%= name %> => <%= name %>.id == id);
-    Object.assign(<%= name %>, data);
-
-    // Put the updated record on memory.
-    <%= pluralizedName %>.splice(indexOf<%= ucName %>ToUpdate, 1, <%= name %>);
-
-    return cb(null, <%= name %>);
+    delete data.id;
+
+    <%= ucName %>.findByIdAndUpdate(id, data, { new: true }) // Using { new: true } to return the modified document rather than the original.
```
Comment (matiasbeckerle) [inline cli/generators/crud/templates/service.js]: What do you think now @AlphaGit? Feel free to keep pushing ideas!

File: cli/generators/crud/templates/service.js (hunk 4 above)
Comment (AlphaGit) [inline cli/generators/crud/templates/service.js]: Thanks! 👍

File: cli/generators/crud/templates/service.js (hunk 6 above)
Comment (AlphaGit) [inline cli/generators/crud/templates/service.js]: Looks great.

After our discussion about the mongoose signature, that comment is definitely needed.

Comment (matiasbeckerle) [inline cli/generators/crud/templates/service.js]: Hahaaha right?

File: api/user/user.service.js [hunk 7]
```diff
@@ -1,72 +1,43 @@
-import moment from 'moment';
-
-const users = [];
+import mongodbConnection from './../database/mongodb';
```
Comment (juampick) [inline api/user/user.service.js]: I think this import is not used. Am I right @matiasbeckerle ?

Comment (matiasbeckerle) [inline api/user/user.service.js]: You're right! Will remove from both places, thanks!

Comment (matiasbeckerle) [inline api/user/user.service.js]: Done in https://github.com/MakingSense/mern-seed/pull/18/commits/df45a349c21aad7b29d5e9db43056fcf17c6e6e4.

Comment (juampick) [inline api/user/user.service.js]: Thanks @matiasbeckerle

**Review by juampick — APPROVED** [review]
Great and hard work here @matiasbeckerle thanks a lot!
